File: snobot_sim_java/java/snobot_sim_java/com/snobot/simulator/motor_sim/GravityLoadDcMotorSim.java
package com.snobot.simulator.motor_sim;

public class GravityLoadDcMotorSim extends BaseDcMotorSimulator
{
    protected final static double sGRAVITY = 9.8;

    protected final GravityLoadMotorSimulationConfig mConfig;

    public GravityLoadDcMotorSim(DcMotorModel aModel, GravityLoadMotorSimulationConfig aConfig)
    {
        super(aModel);

        mConfig = aConfig;
    }


    @Override
    public void update(double cycleTime)
    {
        double extraAcceleration = -sGRAVITY;

        mMotorModel.step(mVoltagePercentage * 12, mConfig.mLoad, extraAcceleration, cycleTime);
    }

    public GravityLoadMotorSimulationConfig getConfig()
    {
        return mConfig;
    }
}

File: snobot_sim/src/main/native/cpp/SnobotSim/MotorSim/GravityLoadDcMotorSim.cpp
/*
 * GravityLoadDcMotorSim.cpp
 *
 *  Created on: May 9, 2017
 *      Author: preiniger
 */

#include "SnobotSim/MotorSim/GravityLoadDcMotorSim.h"

const double GravityLoadDcMotorSim::sGRAVITY = 9.8;

GravityLoadDcMotorSim::GravityLoadDcMotorSim(const DcMotorModel& aMotorModel, const GravityLoadMotorSimulationConfig& config) :
        GravityLoadDcMotorSim(aMotorModel, config.mLoad)
{
}

GravityLoadDcMotorSim::GravityLoadDcMotorSim(const DcMotorModel& aMotorModel, double aLoad) :
        BaseDcMotorSimulator(GetType(), aMotorModel),
        mLoad(aLoad)
{
}

GravityLoadDcMotorSim::~GravityLoadDcMotorSim()
{
}

double GravityLoadDcMotorSim::GetLoad()
{
    return mLoad;
}

void GravityLoadDcMotorSim::Update(double cycleTime)
{
    double extraAcceleration = -sGRAVITY;

    mMotorModel.Step(mVoltagePercentage * 12, mLoad, extraAcceleration, cycleTime);
}
